### ISSUE TITLE
logging doesn't work when using a c++ class header-only approach

### ISSUE THREAD

**Issue by chmorgan** [zephyrproject-rtos/zephyr#79202]
I've been using header-only C++ for a while now, it's convenient not having to split source and header apart. However this isn't compatible with the Zephyr logging system.

For example one cannot do something in example.h like:

```c++
class Example {
private:
    LOG_MODULE_REGISTER(Example, LOG_LEVEL_DBG);

public:
    void some_method()
    {
        LOG_INF("example log info");
    }
};
```

or even putting LOG_MODULE_REGISTER() outside of the class doesn't work either (as example.h may be included in several locations, so the LOG_MODULE_REGISTER() is duplicated several times).

I didn't see any examples that did this with the logging system, and its been a pain separating out .h/.cpp just for the logging system (I'm used to espressif's esp-idf where the logger module is a string passed to each log call and no LOG_MODULE_REGISTER() like thing is necessary. esp-idf logging works great even if used in header only c++ projects).

Thoughts? Maybe I'm missing something relative to advanced usage?

**Comment by dkalowsk** [comment]
I don't think this is a bug.  `LOG_MODULE_REGISTER` needs to be in a C/CPP file.  I think what you want is the `LOG_MODULE_DECLARE`.  Moving to a discussion.
